packages/webview_flutter/webview_flutter_android/lib/src/v4/src/android_webview_controller.dart

```diff
@@ -0,0 +1,320 @@
+import 'android_navigation_delegate.dart';
+import 'android_proxy.dart';
+
+/// Object specifying creation parameters for creating a [AndroidWebViewController].
+///
+/// When adding additional fields make sure they can be null or have a default
+/// value to avoid breaking changes. See [PlatformWebViewControllerCreationParams] for
+/// more information.
+@immutable
+class AndroidWebViewControllerCreationParams
+    extends PlatformWebViewControllerCreationParams {
+  /// Creates a new [AndroidWebViewControllerCreationParams] instance.
+  AndroidWebViewControllerCreationParams({
+    @visibleForTesting this.androidWebViewProxy = const AndroidWebViewProxy(),
+    @visibleForTesting android_webview.WebStorage? androidWebStorage,
+  })  : androidWebStorage =
+            androidWebStorage ?? android_webview.WebStorage.instance,
+        super();
+
+  /// Creates a [AndroidWebViewControllerCreationParams] instance based on [PlatformWebViewControllerCreationParams].
+  factory AndroidWebViewControllerCreationParams.fromPlatformWebViewControllerCreationParams(
+    // Recommended placeholder to prevent being broken by platform interface.
+    // ignore: avoid_unused_constructor_parameters
+    PlatformWebViewControllerCreationParams params, {
+    @visibleForTesting
+        AndroidWebViewProxy androidWebViewProxy = const AndroidWebViewProxy(),
+    @visibleForTesting android_webview.WebStorage? androidWebStorage,
+  }) {
+    return AndroidWebViewControllerCreationParams(
+      androidWebViewProxy: androidWebViewProxy,
+      androidWebStorage:
+          androidWebStorage ?? android_webview.WebStorage.instance,
+    );
+  }
+
+  /// Handles constructing objects and calling static methods for the Android WebView
+  /// native library.
+  @visibleForTesting
+  final AndroidWebViewProxy androidWebViewProxy;
+
+  /// Manages the JavaScript storage APIs provided by the [android_webview.WebView].
+  @visibleForTesting
+  final android_webview.WebStorage androidWebStorage;
+}
+
+/// Implementation of the [PlatformWebViewController] with the Android WebView API.
+class AndroidWebViewController extends PlatformWebViewController {
+  /// Creates a new [AndroidWebViewCookieManager].
+  AndroidWebViewController(PlatformWebViewControllerCreationParams params)
+      : super.implementation(params is AndroidWebViewControllerCreationParams
+            ? params
+            : AndroidWebViewControllerCreationParams
+                .fromPlatformWebViewControllerCreationParams(params));
+
+  AndroidWebViewControllerCreationParams get _androidWebViewParams =>
+      params as AndroidWebViewControllerCreationParams;
+
+  /// The native [android_webview.WebView] being controlled.
+  late final android_webview.WebView _webView = withWeakRefenceTo(this, (
+    WeakReference<AndroidWebViewController> weakReference,
+  ) {
+    return _androidWebViewParams.androidWebViewProxy.createAndroidWebView(
+      useHybridComposition: true,
```

Comment:
We should fix https://github.com/flutter/flutter/issues/108106 as part of v4; the logic should be that useHybridComposition is off by default, and can be enabled (with enabling it being fixed to use `initExpensiveAndroidView` so that it actually does something again).

Reply:
Fixed. Please have a careful look at this however, thinking is not knowing.

Reply:
Capturing from offline discussion: we'll leave this as `true` for now and just not expose it publicly, since the switching logic still doesn't actually work (since it's the same as the current, broken-by-3.0 switching logic). That won't actually cause us to use HC, but will bypass some VD-only cruft.

Later on we can fix the internals, and then plumb `useHybridComposition` through publicly as a creation parameter to allow choosing HC instead of TLHC. We'll make the default `false` at that point instead.